// @api-platform/api-doc-parser/lib/swagger/handleJson.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.removeTrailingSlash = undefined;

var _keys = require("babel-runtime/core-js/object/keys");

var _keys2 = _interopRequireDefault(_keys);

exports.default = function (response, entrypointUrl) {
  var paths = (0, _lodash.uniq)((0, _keys2.default)(response.paths).map(function (item) {
    return item.replace("/{id}", "");
  }));

  var resources = paths.map(function (item) {
    var name = item.replace("/", "");
    var url = removeTrailingSlash(entrypointUrl) + item;
    var firstMethod = (0, _keys2.default)(response.paths[item])[0];
    var title = response.paths[item][firstMethod]["tags"][0];
    var fieldNames = (0, _keys2.default)(response.definitions[title].properties);

    var fields = fieldNames.map(function (fieldName) {
      return new _Field2.default(fieldName, {
        id: null,
        range: null,
        reference: null,
        required: !!response.definitions[title].required.find(function (value) {
          return value === fieldName;
        }),
        description: (0, _lodash.get)(response.definitions[title].properties[fieldName], "description", "")
      });
    });

    return new _Resource2.default(name, url, {
      id: null,
      title: title,
      fields: fields,
      readableFields: fields,
      writableFields: fields
    });
  });

  return resources;
};

var _lodash = require("lodash");

var _Field = require("../Field");

var _Field2 = _interopRequireDefault(_Field);

var _Resource = require("../Resource");

var _Resource2 = _interopRequireDefault(_Resource);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var removeTrailingSlash = exports.removeTrailingSlash = function removeTrailingSlash(url) {
  if (/\/$/.test(url)) {
    url = url.slice(0, -1);
  }
  return url;
};
