// @api-platform/admin/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  AdminBuilder: true,
  Create: true,
  Edit: true,
  fieldFactory: true,
  inputFactory: true,
  resourceFactory: true,
  List: true,
  Show: true
};
Object.defineProperty(exports, "AdminBuilder", {
  enumerable: true,
  get: function get() {
    return _AdminBuilder2.default;
  }
});
Object.defineProperty(exports, "Create", {
  enumerable: true,
  get: function get() {
    return _Create2.default;
  }
});
Object.defineProperty(exports, "Edit", {
  enumerable: true,
  get: function get() {
    return _Edit2.default;
  }
});
Object.defineProperty(exports, "fieldFactory", {
  enumerable: true,
  get: function get() {
    return _fieldFactory2.default;
  }
});
Object.defineProperty(exports, "inputFactory", {
  enumerable: true,
  get: function get() {
    return _inputFactory2.default;
  }
});
Object.defineProperty(exports, "resourceFactory", {
  enumerable: true,
  get: function get() {
    return _resourceFactory2.default;
  }
});
Object.defineProperty(exports, "List", {
  enumerable: true,
  get: function get() {
    return _List2.default;
  }
});
Object.defineProperty(exports, "Show", {
  enumerable: true,
  get: function get() {
    return _Show2.default;
  }
});

var _AdminBuilder2 = _interopRequireDefault(require("./AdminBuilder"));

var _Create2 = _interopRequireDefault(require("./Create"));

var _Edit2 = _interopRequireDefault(require("./Edit"));

var _fieldFactory2 = _interopRequireDefault(require("./fieldFactory"));

var _inputFactory2 = _interopRequireDefault(require("./inputFactory"));

var _resourceFactory2 = _interopRequireDefault(require("./resourceFactory"));

var _List2 = _interopRequireDefault(require("./List"));

var _Show2 = _interopRequireDefault(require("./Show"));

var _hydra = require("./hydra");

Object.keys(_hydra).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _hydra[key];
    }
  });
});

var _docsUtils = require("./docsUtils");

Object.keys(_docsUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _docsUtils[key];
    }
  });
});
